zkvm/src/constraints.rs

```diff
@@ -44,6 +45,42 @@ pub struct CommitmentWitness {
     blinding: Scalar,
 }
 
+impl Constraint {
+    pub fn verify<CS: r1cs::ConstraintSystem>(self, cs: &mut CS) -> Result<(), VMError> {
+        cs.specify_randomized_constraints(move |cs| {
+            let z = cs.challenge_scalar(b"verify challenge");
+
+            // Flatten the constraint into one expression
+            // Note: cloning because we can't move out of captured variable in an `Fn` closure
+            let expr = self.clone().flatten(cs, z);
+
+            // Add the resulting expression to the constraint system
+            cs.constrain(expr.to_r1cs_lc());
+
+            Ok(())
+        })
+        .map_err(|e| VMError::R1CSError(e))
+    }
+
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS, z: Scalar) -> Expression {
```

Comment:
To avoid confusion, i'd flatten into the r1cs linearcombination type and used `zkvm::Expression` only for the operations that are defined in the spec.